File: shoring/compat/src/test/java/com/panopset/compat/test/LogopTest.java
package com.panopset.compat.test;

import java.io.IOException;
import java.util.logging.Level;
import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import com.panopset.compat.Fileop;
import com.panopset.compat.Logop;
import com.panopset.compat.Logop.Listener;
import com.panopset.compat.Logop.LogEntry;
import com.panopset.compat.Stringop;

public class LogopTest {

  LogEntry lr;

  final Listener listener = new Listener() {
    @Override
    public void log(LogEntry logEntry) {
      lr = logEntry;
    }
  };
  
  @Test
  void simpleTest() {
    Logop.clear();
    Logop.info(Stringop.FOO);
    Assertions.assertEquals(1, Logop.getStack().size());
    Logop.info(Stringop.BAR);
  }

  @Test
  void test() throws IOException {
    Logop.clear();
    Logop.getLogger().info(Stringop.FOO);
    Fileop.touch(FileopTest.tempFile);
    Logop.error(Stringop.FOO, FileopTest.tempFile);
    Assertions.assertTrue(Logop.getStack().size() > 0);
    FileopTest.cleanup();
    Logop.clear();
    Logop.addLogListener(listener);
    Logop.dspmsg(Stringop.FOO);
    Assertions.assertEquals(Stringop.FOO, lr.getMessage());
    Assertions.assertEquals(Level.INFO, lr.getLevel());
    Logop.handle(new Exception(Stringop.FOO));
    Assertions.assertEquals(Level.SEVERE, lr.getLevel());
    Logop.debug(Stringop.FOO);
    Assertions.assertEquals(Level.FINE, lr.getLevel());
    Logop.warn(Stringop.FOO);
    Assertions.assertEquals(Level.WARNING, lr.getLevel());
    Logop.turnOnDebugging();
    Logop.warn(Stringop.FOO);
    Assertions.assertEquals(Stringop.FOO, lr.getMessage());
    Assertions.assertEquals(Level.WARNING, lr.getLevel());
    Exception ex = new Exception(Stringop.BAR);
    Logop.error(Stringop.FOO, ex);
    Assertions.assertEquals(7, Logop.getStack().size());
    MatcherAssert.assertThat(12, Matchers.lessThan(Logop.getEntryStackAsText().length()));
    Logop.clear();
    String stackTrace = Logop.getStackTraceAndCauses(ex);
    MatcherAssert.assertThat(12, Matchers.lessThan(stackTrace.length()));
    Logop.clear();
    Exception ex2 = new Exception(ex);
    stackTrace = Logop.getStackTraceAndCauses(ex2);
    MatcherAssert.assertThat(12, Matchers.lessThan(stackTrace.length()));
    Assertions.assertFalse(Logop.isDebugging());
    Logop.turnOnDebugging();
    Assertions.assertTrue(Logop.isDebugging());
    Logop.clear();
    Assertions.assertFalse(Logop.isDebugging());
    for (int i=0; i<1001; i++) {
      Logop.dspmsg(String.format("%s:%d", Stringop.FOO, i));
    }
    Assertions.assertEquals(901, Logop.getStack().size());
    Logop.clear();
    Logop.debug(Stringop.FOO);
    Assertions.assertEquals(1, Logop.getStack().size());
  }

}

File: code/shoring/compat/src/test/kotlin/com/panopset/compat/test/LogopTest.kt
package com.panopset.compat.test

import com.panopset.compat.*
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.logging.Level

class LogopTest {

    @Test
    fun simpleTest() {
        Logz.clear()
        Logz.info(Stringop.FOO)
        Assertions.assertEquals(1, stacklg.size)
        Logz.info(Stringop.BAR)
    }

    @Test
    fun test() {
        Logz.clear()
        Logz.info(Stringop.FOO)
        Fileop.touch(tempFile)
        Logz.errorMsg(Stringop.FOO, tempFile)
        Assertions.assertTrue(stacklg.size > 0)
        FileopTest.cleanup()
        Logz.clear()
        Logz.dspmsg(Stringop.FOO)
        Assertions.assertEquals(Stringop.FOO, stacklg.peek().message)
        Assertions.assertEquals(Level.INFO, stacklg.peek().level)
        Logz.errorEx(Exception(Stringop.FOO))
        Assertions.assertEquals(Level.SEVERE, stacklg.peek().level)
        Logz.debug(Stringop.FOO)
        Assertions.assertEquals(Level.FINE, stacklg.peek().level)
        Logz.warn(Stringop.FOO)
        Assertions.assertEquals(Level.WARNING, stacklg.peek().level)
        Logz.warn(Stringop.FOO)
        Assertions.assertEquals(Stringop.FOO, stacklg.peek().message)
        Assertions.assertEquals(Level.WARNING, stacklg.peek().level)
        val ex = Exception(Stringop.BAR)
        Logz.errorMsg(Stringop.FOO, ex)
        Assertions.assertEquals(7, stacklg.size)
        Assertions.assertTrue(12 < getEntryStackAsTextlg().length)
        Logz.clear()
        var stackTrace = getStackTraceAndCauseslg(ex)
        Assertions.assertTrue(12 < stackTrace.length)
        Logz.clear()
        val ex2 = Exception(ex)
        stackTrace = getStackTraceAndCauseslg(ex2)
        Assertions.assertTrue(12 < stackTrace.length)
        Logz.clear()
        for (i in 0..100) {
            Logz.dspmsg(String.format("%s:%d", Stringop.FOO, i))
        }
        Assertions.assertEquals(11, stacklg.size)
        Logz.clear()
        Logz.debug(Stringop.FOO)
        Assertions.assertEquals(1, stacklg.size)
    }
}
